### DISCUSSION TITLE
Segmented columns of table spanning multiple pages

### DISCUSSION THREAD

**Discussion by pdille** [jsvine/pdfplumber#864] [Ask for help with specific PDFs]
I will preface by saying I do not have control over the format of these pdfs, heh. Also apologies if the title isn't clear, hopefully discussion below will clarify.

This all may be a pipe dream, but I wanted to see what's possible to simplify the extraction of data from pdfs that have tables that look like the following examples, most specifically the second one.

In this first example

[20230306-test.pdf](https://github.com/jsvine/pdfplumber/files/11228780/20230306-test.pdf)

We see that columns 5-8 of the second page have more rows than the other columns and unfortunately the cells of empty data don't have border boundaries to them. I will note that pdfplumber handles this pdf just fine and the data it extracts for the last few rows are of the form:

`[None, None, None, None, '23:45', '63', '0', '0', None, None, None, None, None, None, None, None, None]`

which is great in that it kept the content to the correct column positions. No other python pdf parser I tried did this correctly and all others just stuck the data it found into the first 4 columns, which is not right. So that's another win for pdfplumber. :-)

No problems so far, but I include that one as an intro into a more complicated case.  

In this next pdf ([20230411-test.pdf](https://github.com/jsvine/pdfplumber/files/11228781/20230411-test.pdf)), note on the last page we have a similar situation where some columns end up with more rows than others. Now unfortunately in this case we have data in the first 4 columns, a large gap and then data corresponding to the last 5 columns. Once again we don't have cell boundaries for the empty ones and here this results in basically 2 separate tables. That's what pdfplumber ends up detecting here with all default settings.  I was semi-successful in using the 'join_tolerance' parameter with a larger value than the default one and pdfplumber then detected a single table with a row extracted as:

`['22:45', '56', '0', '0', '', '23:45', '48', '0.002165', '2.1649757', '']`

That's closer, but what I wish it gave me was:

`['22:45', '56', '0', '0', None, None, None, None, None, None, None, None, '23:45', '48', '0.002165', '2.1649757', '']`

but obviously it could only do that if it knew about the overall table structure from the page that came before it.  In the first pdf I showed above where it handled some columns with more rows than others it at least had reference on the same page of what a full row looks like.  In this second example pdf it doesn't know how many columns there truly should be.  Is it possible for pdfplumber to reference table structure from a prior page?  Or is there a better way going about extracting the info but preserving the columns the data corresponds to?

Thanks again for such a great library and much appreciate again for your help in my other discussion on here.

**Comment by jsvine** [comment]
That's a tricky one. My advice would be to work with the `Table` objects directly via `page.find_tables(...)`, and then to apply your own logic,  i.e.:

```python
tables = page.find_tables(...)
for table in tables:
  for bbox in tables.cells:
    cell_text = page.crop(bbox).extract_text()
    [...]
```

Another approach might be to use information from the previous page to construct a set of explicit horizontal and vertical lines.

**Comment by pdille** [comment]
Thanks for the reply! In regards to your second suggestion, what's the most efficient way of getting the coordinates of the vertical lines of a table? As a quick manual test I used the following code:

```
page_info = page.debug_tablefinder(table_settings={})
print(page_info.cells)
```
And then this gave me a list of 4-tuples of what I believe are the bounding boxes (x0, top, x1, bottom) of all the cells of all the tables on the page.  I then manually (though I can certainly automate this) pulled out all the unique x0, x1 coordinates and populated a list to be passed to `explicit_vertical_lines`.  I also made use of your `explicit_horizontal_line` suggestion of page edge tops from another thread we had together.

```
    table_settings = {
      "explicit_horizontal_lines": [ edge["top"] for edge in page.edges ],
      "explicit_vertical_lines": [50.86098, 129.76101, 195.10101, 266.02101, 336.94099500000004, 427.00101, 492.340995, 563.2610399999999,  634.1809800000001, 716.26098, 781.60101, 852.52098, 923,44101, 1004.74101, 1070.08104, 1141.0009799999999, 1211.920979999999999, 1467.82098]
    }

    page_table = page.extract_table(table_settings=table_settings)
    print(page_table)
```

Combining that all into the above, pdfplumber extracted rows from the last page of [20230411-test.pdf](https://github.com/jsvine/pdfplumber/files/11228781/20230411-test.pdf) in the column format I was hoping for!

`['22:00', '56', '0.26062', '260.62975', '', '', '', '', '', '', '', '', '22:30', '48', '0.0002316', '0.2316402', '']`
(Empty strings or Nones are perfectly fine for empty cells)

So despite the large gap in columns described in the first post, I was able to use the table structure info from the prior page to ensure I got the same number of cells being extracted, with data in the correct column positions. Huzzah. Heh.  

Thanks for pointing me in this direction. Is there a better way though to extract out the vertical coordinates than using the debug_tablefinder I used above?  And is this what you had in mind when you were thinking of when you suggested this path to take?

Thanks!

**Reply by jsvine** [reply]
Nice! Re. this:

> Thanks for pointing me in this direction. Is there a better way though to extract out the vertical coordinates than using the debug_tablefinder I used above? 

If you're confident that the first row of the "canonical" table has all the necessary cells, then you could try this:

```python
table = page.find_tables(...)[0]
first_row_cells = table.rows[0].cells
vertical_positions = [ x0 for x0, top, x1, bottom in first_row_cells ]
```

**Reply by pdille** [reply]
Awesome, that's a nice quick way to get all the coordinates I need. And yes for this swath of pdfs I have, the logic to assume the first row has all the cells is justified based on the format I'm dealing with.

Is there a way though to find only the largest table on the page?  I know with extract_table (singular) it finds the largest table and extracts data from that, but is there a way to have find_tables (or perhaps new method called find_table with a specific flag passed in) to return only the largest table object?  I assume some logic already exists that is being used for extract_table.

Right now with the format of the pdfs I'm dealing with I can always assume the last table on the first page (to be used to get the vertical positions for subsequent pages) is the one I want, but would be nice to not always rely on that.

Thanks again for your time answering my questions on here!

**Reply by jsvine** [reply]
> Is there a way though to find only the largest table on the page?

Thanks! I think you've identified a "missing" method for `pdfplumber.Page`. I've opened an issue to address it: https://github.com/jsvine/pdfplumber/issues/873

... but probably won't have time to incorporate it immediately.

In the meantime, you could just copy over the logic that `Page.extract_table(...)` uses, while skipping the final extraction step:

https://github.com/jsvine/pdfplumber/blob/ae676aeacd958e7b1572f35568e612314d611eff/pdfplumber/page.py#L303-L318

**Reply by pdille** [reply]
Thanks! Appreciate the pointer to the code and willingness to add this as an eventual new feature!
